cmake/IBAMRTesting.cmake

```diff
@@ -0,0 +1,285 @@
+# Test configuration
+set(IBAMR_TEST_DIR "${CMAKE_SOURCE_DIR}/tests/" CACHE PATH "Location of the test cases")
+set(IBAMR_TEST_TIMEOUT 600 CACHE STRING "Test timeout")
+set(IBAMR_MAX_PARALLEL CACHE STRING 1 "")
+set(IBAMR_TEST_CLEANUP FALSE CACHE BOOL "")
+
+# Find numdiff
+find_program(NUMDIFF_EXECUTABLE NAMES numdiff HINTS ${NUMDIFF_ROOT} PATH_SUFFIXES bin)
+if ("${NUMDIFF_EXECUTABLE}" STREQUAL "NUMDIFF_EXECUTABLE-NOTFOUND")
+  message(WARNING "\
+The configuration script was not able to locate numdiff. If you want to run \
+the test suite you will need to either edit attest.conf, specify the path to \
+numdiff to attest, or rerun CMake with the argument NUMDIFF_ROOT specifying \
+numdiff's root installation directory.")
+  # clear the value so that attest.conf doesn't contain an invalid path
+  set(NUMDIFF_EXECUTABLE "")
+  set(IBAMR_ENABLE_TESTING FALSE)
+endif ()
+
+set(IBAMR_ALLOW_OVERSUBSCRIBED OFF "Allow tests to be added if they are oversubscribed")
+mark_as_advanced(
+  NUMDIFF_EXECUTABLE
+  IBAMR_TEST_DIR
+  IBAMR_TEST_TIMEOUT
+  IBAMR_ALLOW_OVERSUBSCRIBED)
+
+function (_ibamr_parse_test name)
+  set(n_mpirun 1)
+  set(n_restart 0)
+  set(expect_error FALSE)
+
+  string(REPLACE "." ";" name_list "${name}")
+  foreach (override ${name_list})
+    string(REPLACE "=" ";" _x ${override})
+    list(LENGTH _x _x_len)
+    if (${_x_len} EQUAL 2)
+      list(GET _x 0 _x_key)
+      list(GET _x 1 _x_value)
+      if (_x_key STREQUAL "mpirun")
+        set(n_mpirun ${_x_value})
+      elseif (_x_key STREQUAL "expect_error")
+        string(TOUPPER "${_x_value}" value)
+        set("expect_error" ${value})
+      elseif (_x_key STREQUAL "restart")
+        set("restart" ${_x_value})
+      else (_x_key STREQUAL "restart")
+        message(WARNING "Encountered unrecognized override ${_x_key}")
+      endif ()
+      message(STATUS "${_x_key} ${_x_value}")
+    endif ()
+  endforeach ()
+
+  set(n_mpirun ${n_mpirun} PARENT_SCOPE)
+  set(n_restart ${n_restart} PARENT_SCOPE)
+  set(expect_error ${expect_error} PARENT_SCOPE)
+endfunction()
+
+#[[!md
+
+ibamr_add_testss(
+  base_name
+  DIRECTORY
+    TestDirectory
+)
+
+]]#
+function (ibamr_add_tests name)
+  cmake_parse_arguments(test
+    ""
+    "DIRECTORY"
+    "LIBRARIES;SOURCES" ${ARGN})
+
+  if (NOT test_DIRECTORY)
+    message(FATAL_ERROR "Missing DIRECTORY for test")
+  endif ()
+
+  if (NOT IBAMR_ENABLE_TESTING)
+    return ()
+  endif ()
+
+  #message(CHECK_START "Adding tests: ${test_DIRECTORY}/${name}")
+  #list(APPEND CMAKE_MESSAGE_INDENT "  ")
+
+  # Add test source if it exists (may not because of _2d/_3d suffix)
+  if (EXISTS "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
+    list(APPEND test_SOURCES "${IBAMR_TEST_DIR}/${test_DIRECTORY}/${name}.cpp")
```

Comment:
I am not sure what you mean my "automake-like names".

The "SOURCES" is there to make it easier for me to use the same CMake backend for 2d/3d/manual tests. In general, the sources should not required when adding a test and will report an error if the driver function is not detected.

Reply:
by automake-like names I mean `test_SOURCES`, since that's the convention used with autotools' automake files:
https://github.com/IBAMR/IBAMR/blob/3922a88d91bbc9461aea61c76160455f4cfb6bbd/tests/IBFE/Makefile.am#L32

> In general, the sources should not required when adding a test and will report an error if the driver function is not detected.

Oh, I see - this is set up in a different way than I originally thought. I thought `test_SOURCES` was specific to exactly one executable.